src/model/keychain.js

```diff
@@ -136,3 +143,86 @@ export async function hasKey(key) {
   }
   return null;
 }
+
+export async function restoreBackupIntoKeychain(backedUpData) {
+  // Access control config per each type of key
+  const publicAccessControlOptions = {
+    accessible: ACCESSIBLE.ALWAYS_THIS_DEVICE_ONLY,
+  };
+
+  let privateAccessControlOptions = {};
+  const canAuthenticate = await canImplyAuthentication({
+    authenticationType: AUTHENTICATION_TYPE.DEVICE_PASSCODE_OR_BIOMETRICS,
+  });
+
+  let isSimulator = false;
+
+  if (canAuthenticate) {
+    isSimulator = __DEV__ && (await DeviceInfo.isEmulator());
+  }
+  if (canAuthenticate && !isSimulator) {
+    privateAccessControlOptions = {
+      accessControl: ACCESS_CONTROL.USER_PRESENCE,
+      accessible: ACCESSIBLE.WHEN_UNLOCKED_THIS_DEVICE_ONLY,
+    };
+  }
+
+  Object.keys(backedUpData).forEach(async key => {
+    const value = backedUpData[key];
+    let accessControl = publicAccessControlOptions;
+    if (
+      (key.indexOf('rainbowSeedPhrase') !== -1 ||
+        key.indexOf('rainbowPrivateKey') !== -1) &&
+      key !== 'rainbowSeedPhraseMigratedKey'
+    ) {
+      accessControl = privateAccessControlOptions;
+    }
+    if (typeof value === 'string') {
+      await saveString(key, value, accessControl);
+    } else {
+      await saveObject(key, value, accessControl);
+    }
+  });
+
+  // Save the migration flag
+  // to prevent this flow in the future
+  await saveString(
+    'rainbowSeedPhraseMigratedKey',
+    'true',
+    publicAccessControlOptions
+  );
+
+  return true;
+}
+
+// Attempts to save the password to decrypt the backup from the iCloud keychain
+export async function saveBackupPassword(password) {
+  try {
+    await setSharedWebCredentials('rainbow.me', 'Backup Password', password);
+  } catch (e) {
+    logger.log('Error while backing up password', e);
+  }
+}
+
+// Attempts to fetch the password to decrypt the backup from the iCloud keychain
+export async function fetchBackupPassword() {
+  try {
+    const results = await requestSharedWebCredentials();
+    return results.password;
+  } catch (e) {
+    logger.log('Error while fetching backup password', e);
```

Comment:
Just a note: we have a return value for the success case but not for the catch. Bc JS, it will still return undefined so we are technically safe here. I think in other places we have been explicitly returning null.

Reply:
Will improve once we have typescript support